### DISCUSSION TITLE
[New extension] Project lpp: Introduces OOP to Scratch!

### DISCUSSION THREAD

**Discussion by FurryR** [TurboWarp/extensions#1228] [Show and tell]
<div align="center">

# 🐺 [lpp](https://github.com/FurryR/lpp-scratch)

> A high-level programming language based on Scratch developed by @FurryR.

</div>

<table>
<tr><td>

## 🛠️ Features

- ✅ Use lambda functions (functional programming) like @LilyMakesThings's `inline` but more powerful and works fine on compiler mode.
- ✅ Make complex objects like class instances.
- ✅ Visual debugger & traceback.
- ✅ Direct JSON construction without `JSON.parse` in Scratch.
- ✅ Exceptions & try-catch system.
- ✅ Use your favorite Promise and stackless coroutine (`yield`, `yield*`, `await`).
- ✅ Fully compatible with original Scratch (using [`Eureka`](https://github.com/EurekaScratch/eureka-loader)), even with early versions (before commit `39b18fe` by @mzgoddard, Apr 12, 2019).
- ✅ `zh-cn`, `ja-jp` (WIP) and `en-us` translation + `en-us` documentation.

<div align="center">

🤖 [See more detail & try out demo](https://github.com/FurryR/lpp-scratch)

</div>
</td></tr>
<tr><td>

## 🤔 Compatibility

| 🎮 Platform | 👾 Availability | ❓ Tested? |
| --- | --- | --- |
| [Turbowarp](https://turbowarp.org) | ✅ (Preferred) | ✅ |
| [Gandi cocrea](https://cocrea.world) | ✅ | ✅ |
| [Gandi Co-create world](https://ccw.site/) | ✅ | ✅ |
| [Original Scratch](https://scratch.mit.edu) | ✅ (Eureka) | ✅ |
| [Xueersi](https://code.xueersi.com/) | ✅ (Eureka, early Scratch) | ✅ |
| [ClipCC](https://codingclip.com) | ✅ (Eureka) | ✅ |
| [aerfaying](https://aerfaying.com) | ✅ (Eureka) | ✅ |
| [PenguinMod](https://penguinmod.com) | ❗ (weird UI) | ✅ |
| [40code](https://www.40code.com) | ✅ (Eureka) | ❌ |

</td></tr>
<tr><td>

## 🚧 Progress

- ✅ Lambda function (arguments, closure)
- ✅ Prototype, `new`, `self`
- ✅ Control flow (Return, Exceptions & try-catch)
- ✅ Direct JSON construction & (de)serialization (`JSON` namespace)
- ✅ Local variables & scope
- ✅ Visual debugger
- 🕘 Builtin types (`String`, `Boolean`, etc.)
- 🕘 Builtin utility (`JSON`, `Math`, etc.)
- 🕘 Operators
- ❌ Documentation
- ❌ Promise, generator & async functions
- ❌ Some additional syntactic sugar (`rest` operator, etc.)

💞 Open for collaborations! [Tell me](https://discordapp.com/users/im_furryr) if you want to join.

</td></tr>
</table>

**Comment by lego7set** [comment]
i bet this took a lot of reading and thinking to do. your extension looks awesome.

**Comment by ExpertNoob1019** [comment]
THANKS SO MUCH I NEEDED THIS!

**Comment by milkyestway** [comment]
seems promising, please continue posting updates and if you run into a problem, maybe we could help you!

**Comment by penta-quark-neutro** [comment]
How can I add tooltips to my extensions?
I tried to figure out how your extension does it, but I couldn't quite understand how.

**Comment by FurryR** [comment]
Kinda old-school fashioned but unfortunately the project is no longer maintained. btw, it's totally out of AI.

**Comment by penta-quark-neutro** [comment]
Yes, I had already noticed that it isn't being maintained, and I don't understand the mention of AI.
But I'm only interested in the tooltips so I can add them to my extensions; I just couldn't figure out how you implemented that.

**Reply by FurryR** [reply]
> Yes, I had already noticed that it isn't being maintained, and I don't understand the mention of AI. But I'm only interested in the tooltips so I can add them to my extensions; I just couldn't figure out how you implemented that.

Patch cbfsb (`convertBlocksForScratchBlocks`? or something else) and add a `tooltip` to `tooltip` property and you are okay to go.

**Reply by penta-quark-neutro** [reply]
I appreciate the information. Thank you.

**Comment by penta-quark-neutro** [comment]
It seems I didn't get anywhere.
I ran into a problem: "vm error Failed to register primitives for extension on service unsandboxed.0.exps: TypeError: Cannot assign to read only property 'exps_me0' of object '#<Object>'".
Partly because I couldn't find "convertBlocksForScratchBlocks".
I found a few things, but they didn't help at all.

**Reply by FurryR** [reply]
Check LMT's or SharkPool's extensions and you will find out.
